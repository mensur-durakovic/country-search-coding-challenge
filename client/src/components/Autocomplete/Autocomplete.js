import React, { Component } from "react";
import PropTypes from "prop-types";
import styles from "./Autocomplete.module.css";
import { fetchCountries } from "../../api/api";

const UP_ARROW_KEY = 38;
const DOWN_ARROW_KEY = 40;
const ENTER_KEY = 13;
const ESCAPE_KEY = 27;

export default class Autocomplete extends Component {
  static propTypes = {
    labelText: PropTypes.string,
    noSuggestionsText: PropTypes.string,
  };

  static defaultProps = {
    labelText: "Type something",
    noSuggestionsText: "No suggestions",
  };

  state = {
    loadingSuggestions: false,
    errorSuggestions: false,
    filteredSuggestions: [],
    currentSuggestionIndex: 0,
    displaySuggestions: false,
    enteredText: "",
    suggestionClicked: false
  };

  constructor(props) {
    super(props);
    this.suggestionsListRef = React.createRef();
    this.searchInputRef = React.createRef();
    this.timer = null;
  }

  componentDidMount() {
    //this.fetchSuggestionsData();
    document.addEventListener("mousedown", this.clickOutsideHandler);
  }

  componentWillUnmount() {
    document.removeEventListener("mousedown", this.clickOutsideHandler);
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.enteredText !== this.state.enteredText && !this.state.suggestionClicked) {
      this.fetchSuggestionsData();
    }
  }

  clickOutsideHandler = (e) => {
    if (
      this.suggestionsListRef &&
      this.suggestionsListRef.current &&
      !this.suggestionsListRef.current.contains(e.target)
    ) {
      this.setState({
        currentSuggestionIndex: 0,
        displaySuggestions: false,
      });
    }
  };

  setInputValue = (e) => {
    this.setState({
      enteredText: e.target.value,
      suggestionClicked: false,
    });
  };

  fetchSuggestionsData = async (e) => {
    try {
      clearTimeout(this.timer);
      const enteredFilter = this.state.enteredText.toLowerCase();

      //fetch data from backend after 0.5 sec after user stops typing, so we don't make unecessarry calls
      this.timer = setTimeout(async () => {
        this.setState({ loadingSuggestions: true });

        const response = await fetchCountries(enteredFilter);

        this.setState({
          loadingSuggestions: false,
          errorSuggestions: false,
          currentSuggestionIndex: 0,
          filteredSuggestions: response.data.countries.map((c) => c.name),
          displaySuggestions: true,
        });

        //focus search input after side effect is finished
        this.searchInputRef.current.focus();

      }, 500);
    } catch (error) {
      console.error("Error on fetching countries", error);
      this.setState({
        errorSuggestions: true,
        loadingSuggestions: false,
      });
    }
  };

  focusSuggestion = () => {
    const { currentSuggestionIndex, filteredSuggestions } = this.state;
    const suggestionId = `suggestionId-${
      filteredSuggestions[currentSuggestionIndex - 1]
    }`;
    const suggestionTarget = document.getElementById(suggestionId);
    if (suggestionTarget) {
      suggestionTarget.scrollIntoView();
    }
  };

  highlightMatchedSubstring = (suggestion) => {
    const { enteredText } = this.state;
    const parts = suggestion.split(new RegExp(`(${enteredText})`, "gi"));
    const result = parts.map((p, i) => (
      <span
        key={i}
        className={
          p.toLowerCase() === enteredText.toLowerCase() ? styles.markedText : ""
        }
      >
        {p}
      </span>
    ));

    return result;
  };

  clickHandler = (e) => {
    this.setState({
      currentSuggestionIndex: 0,
      filteredSuggestions: [],
      displaySuggestions: false,
      enteredText: e.currentTarget.innerText,
      suggestionClicked: true,
    });
  };

  /**
   * keydown handler for better UX
   */
  keyDownHandler = (e) => {
    const { currentSuggestionIndex, filteredSuggestions } = this.state;

    switch (e.keyCode) {
      case ENTER_KEY: {
        this.setState({
          currentSuggestionIndex: 0,
          displaySuggestions: false,
          enteredText: filteredSuggestions[currentSuggestionIndex],
        });
        break;
      }
      case UP_ARROW_KEY: {
        if (currentSuggestionIndex === 0) {
          return;
        }

        this.setState(
          { currentSuggestionIndex: currentSuggestionIndex - 1 },
          this.focusSuggestion
        );

        break;
      }
      case DOWN_ARROW_KEY: {
        if (currentSuggestionIndex + 1 === filteredSuggestions.length) {
          return;
        }
        this.setState(
          { currentSuggestionIndex: currentSuggestionIndex + 1 },
          this.focusSuggestion
        );

        break;
      }
      case ESCAPE_KEY: {
        this.setState({
          currentSuggestionIndex: 0,
          displaySuggestions: false,
        });
        break;
      }
      default:
        break;
    }
  };

  render() {
    const {
      currentSuggestionIndex,
      filteredSuggestions,
      displaySuggestions,
      enteredText,
      loadingSuggestions,
      errorSuggestions,
    } = this.state;

    const { labelText, noSuggestionsText } = this.props;
    const isError = !loadingSuggestions && errorSuggestions;
    const isLoading = loadingSuggestions && !errorSuggestions;

    let suggestionsList;
    if (isError) {
      suggestionsList = (
        <div className={styles.errorFetchingData}>
          <span>{"An error occurred while fetching data!"}</span>
        </div>
      );
    } else if (displaySuggestions && enteredText) {
      if (filteredSuggestions.length) {
        suggestionsList = (
          <ul className={styles.suggestions} ref={this.suggestionsListRef}>
            {filteredSuggestions.map((suggestion, index) => (
              <li
                className={
                  index === currentSuggestionIndex
                    ? styles.activeSuggestion
                    : ""
                }
                id={`suggestionId-${suggestion}`}
                key={suggestion}
                onClick={this.clickHandler}
              >
                {this.highlightMatchedSubstring(suggestion)}
              </li>
            ))}
          </ul>
        );
      } else {
        suggestionsList = (
          <div className={styles.noSuggestions}>
            <span>{noSuggestionsText}</span>
          </div>
        );
      }
    }

    const autoCompleteInput = (
      <>
        <label htmlFor="countriesAutocompleteInput">{labelText}</label>
        <input
          id="countriesAutocompleteInput"
          autoComplete="new-password"
          ref={this.searchInputRef}
          className={styles.input}
          type="text"
          placeholder="type something"
          onChange={this.setInputValue}
          onKeyDown={this.keyDownHandler}
          value={enteredText}
          disabled={isError}
        />
        {suggestionsList}
      </>
    );

    return (
      <>
        {isLoading ? (
          <span className={styles.loadingText}>{"Loading..."}</span>
        ) : (
          autoCompleteInput
        )}
      </>
    );
  }
}
