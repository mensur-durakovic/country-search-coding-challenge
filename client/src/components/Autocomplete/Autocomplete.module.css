input {
    padding: 0.6em;
    width: 20vw;
    font-size: 1em;
    margin: 0 auto;
    border: 0;
    background-color: #f1f1f1;
  }
  
  input:disabled {
    cursor: not-allowed;
  }
  
  label {
    margin-top: 20vh;
    width: 20vw;
    padding: 0.6em;
    text-align: center;
    background-color: #a3d3ff;
    border-radius: 15px;
    margin-bottom: 20px;
  }
  
  .errorFetchingData{
    color: #d63447;
    padding: 0.6em;
    text-align: center;
  }
  
  .suggestions {
    width: calc(20vw + 1rem + 4px);
    max-height: 20vh;
    list-style: none;
    overflow-y: auto;
    margin-top: 0;
    border: 1px solid #d9d9d9;
    border-top-width: 0;
    padding-left: 0;
  }
  
  .suggestions li {
    padding: 0.6rem;
  }
  
  .suggestions li:hover {
    cursor: pointer;
    background-color: #a3d3ff;
  }
  
  .suggestions li:not(:last-of-type) {
    border-bottom: 1px solid #d9d9d9;
  }
  
  .activeSuggestion {
    background-color: #a3d3ff;
  }
  
  .noSuggestions {
    color: #666;
    padding: 0.5rem;
    width: 20vw;
    text-align: center;
  }
  
  .loadingText {
    margin-top: 20vh;
    width: 20vw;
    padding: 0.6em;
    text-align: center;
  }
  
  .markedText {
    font-weight: bold;
  }
  
  @media only screen and (max-width : 1024px) {
    input, label {
      width: 70vw;
    }
    .suggestions {
      width: calc(70vw + 1rem + 4px);
    }
    .noSuggestions {
      width: 70vw;
    }
  }