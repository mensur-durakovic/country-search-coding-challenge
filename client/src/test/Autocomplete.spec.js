import React from 'react';
import renderer from 'react-test-renderer';
import Autocomplete from '../components/Autocomplete/Autocomplete';

test('Autocomplete component', () => {
  const tree = renderer.create(<Autocomplete 
    noSuggestionsText="No suggestions, please try to type something else"
    labelText="Simple autocomplete"/>).toJSON()
  expect(tree).toMatchSnapshot()
})