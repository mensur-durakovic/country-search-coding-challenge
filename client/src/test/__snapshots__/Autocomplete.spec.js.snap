// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete component 1`] = `
Array [
  <label
    htmlFor="countriesAutocompleteInput"
  >
    Simple autocomplete
  </label>,
  <input
    autoComplete="new-password"
    className="input"
    disabled={false}
    id="countriesAutocompleteInput"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="type something"
    type="text"
    value=""
  />,
]
`;
